src/util/WriteOnceSharedVar.h

```diff
@@ -0,0 +1,242 @@
+#pragma once
+
+#include "util/NonCopyable.h"
+#include <exception>
+#include <memory>
+
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+// This is a helper class somewhat like a std::shared_future but with no
+// integration with async / threading / blocking. It's just a sharable,
+// write-once variable you can poll for readiness and store either a value or an
+// exception in. Supports "dataflow"-style programming.
+
+namespace stellar
+{
+template <typename T> class WriteOnceSharedVar
+{
+    struct Impl : public NonMovableOrCopyable
+    {
+        std::unique_ptr<T> mValue;
+        std::exception_ptr mException;
+    };
+
+    std::shared_ptr<Impl> mImpl;
+
+  public:
+    WriteOnceSharedVar() : mImpl(std::make_shared<Impl>())
+    {
+    }
+
+    // Helper to construct in written state.
+    WriteOnceSharedVar(T const& v) : mImpl(std::make_shared<Impl>())
+    {
+        setValue(v);
+    }
+
+    bool
+    hasValue() const
+    {
+        assert(mImpl);
+        return static_cast<bool>(mImpl->mValue);
+    }
+
+    bool
+    hasException() const
+    {
+        assert(mImpl);
+        return static_cast<bool>(mImpl->mException);
+    }
+
+    void
+    setValue(T const& v)
+    {
+        if (hasValue())
+        {
+            throw std::runtime_error("write-once variable value already set");
+        }
+        if (hasException())
+        {
+            throw std::runtime_error(
+                "write-once variable exception already set");
+        }
+        mImpl->mValue = std::make_unique<T>(v);
+    }
+
+    // An "idempotent" set that either sets the value if it's not set,
+    // or asserts that the parameter is equal to the already-set value.
+    //
+    // If an exception is already set, it is left in place and neither
```

Comment:
The behavior when there is an exception is quite weird to me and seems error prone: I would imagine that clobbering an error may not always be intended.

It would be more natural to have the caller reset their shared state explicitely when they want to do this (like when they retry), maybe adding a `reset` or `clear` method would be enough.

Reply:
N/A at this point, dropping this class entirely.